### PR TITLE
Add 'Output.all' combinator for Python

### PR DESCRIPTION
**Pull request by swgillespie**
Output.all is a useful combinator that we already have in Node that
allows the composition of a list of outputs into an output of a list.
This is very useful when authoring components and its lack of presence
in Python was an oversight.

This commit adds 'Output.all' and 'Output.from_input', adding tests and
documentation for each.

Fixes https://github.com/pulumi/pulumi/issues/2284

### CODE REVIEW

File: sdk/python/lib/test/test_next_serialize.py [hunk 1]
```diff
@@ -104,7 +104,7 @@ def tearDown(self):
         self.assertEqual("hello.txt", prop["path"])
 
     @async_test
-    async def test_file_asset(self):
+    async def test_remote_asset(self):
```
Comment (swgillespie) [inline sdk/python/lib/test/test_next_serialize.py]: This fix is unrelated, this test was accidentally named the same as the above test. This works, for esoteric Python reasons, but it confused my code editor.

**Comment by pgavlin** [comment]
It doesn't look like this deals with nested outputs. Should it?

**Comment by swgillespie** [comment]
> It doesn't look like this deals with nested outputs. Should it?

Unclear - do we want it to? We don't have the benefit of the TypeScript type system to make sense of it.

**Comment by swgillespie** [comment]
The most recent commit augments `Output.from_input` to fully unwrap nested inputs, the same way that JavaScript does. There's also a new test for it.

**Comment by swgillespie** [comment]
ping - would like to get this in this week before the holidays.
